waku/v2/peermanager/peer_selection.go

```diff
@@ -30,73 +32,118 @@ func (pm *PeerManager) SelectPeerByContentTopics(proto protocol.ID, contentTopic
 		}
 		pubsubTopics = append(pubsubTopics, pubsubTopic)
 	}
-	return pm.SelectPeer(PeerSelectionCriteria{PubsubTopics: pubsubTopics, Proto: proto, SpecificPeers: specificPeers})
+	peers, err := pm.SelectPeers(PeerSelectionCriteria{PubsubTopics: pubsubTopics, Proto: proto, SpecificPeers: specificPeers})
+	if err != nil {
+		return "", err
+	}
+	return peers[0], nil
 }
 
 // SelectRandomPeer is used to return a random peer that supports a given protocol.
 // If a list of specific peers is passed, the peer will be chosen from that list assuming
 // it supports the chosen protocol, otherwise it will chose a peer from the service slot.
 // If a peer cannot be found in the service slot, a peer will be selected from node peerstore
 // if pubSubTopic is specified, peer is selected from list that support the pubSubTopic
-func (pm *PeerManager) SelectRandomPeer(criteria PeerSelectionCriteria) (peer.ID, error) {
+func (pm *PeerManager) SelectRandom(criteria PeerSelectionCriteria) (peer.IDSlice, error) {
 	// @TODO We need to be more strategic about which peers we dial. Right now we just set one on the service.
 	// Ideally depending on the query and our set  of peers we take a subset of ideal peers.
 	// This will require us to check for various factors such as:
 	//  - which topics they track
 	//  - latency?
-
-	peerID, err := pm.selectServicePeer(criteria.Proto, criteria.PubsubTopics, criteria.Ctx, criteria.SpecificPeers...)
-	if err == nil {
-		return peerID, nil
+	peerIDs, err := pm.selectServicePeer(criteria)
+	if err == nil && len(peerIDs) == criteria.MaxPeers {
+		return maps.Keys(peerIDs), nil
 	} else if !errors.Is(err, ErrNoPeersAvailable) {
 		pm.logger.Debug("could not retrieve random peer from slot", zap.String("protocol", string(criteria.Proto)),
 			zap.Strings("pubsubTopics", criteria.PubsubTopics), zap.Error(err))
-		return "", err
+		return nil, err
+	} else {
+		peerIDs = make(peersMap)
 	}
-
 	// if not found in serviceSlots or proto == WakuRelayIDv200
 	filteredPeers, err := pm.FilterPeersByProto(criteria.SpecificPeers, criteria.Proto)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	if len(criteria.PubsubTopics) > 0 {
 		filteredPeers = pm.host.Peerstore().(wps.WakuPeerstore).PeersByPubSubTopics(criteria.PubsubTopics, filteredPeers...)
 	}
-	return selectRandomPeer(filteredPeers, pm.logger)
+	randomPeers, err := selectRandomPeers(filteredPeers, criteria.MaxPeers-len(peerIDs))
+	if err != nil && len(peerIDs) == 0 {
+		return nil, err
+	}
+
+	for tmpPeer := range randomPeers {
+		peerIDs[tmpPeer] = struct{}{}
+	}
+	return maps.Keys(peerIDs), nil
+}
+
+func getRandom(filter peersMap, count int) (peersMap, error) {
+	i := 0
+	selectedPeers := make(peersMap)
+	for pID := range filter {
+		selectedPeers[pID] = struct{}{}
+		i++
+		if i == count {
+			break
+		}
+	}
+	if len(selectedPeers) == 0 {
+		return nil, ErrNoPeersAvailable
+	}
+	return selectedPeers, nil
 }
```

Comment:
No random function?

Reply:
Map's iterator already has randomness in golang and hence no need to use any additional random function.